### DISCUSSION TITLE
Proprietary software extending RDF4J

### DISCUSSION THREAD

**Discussion by frensjan** [eclipse-rdf4j/rdf4j#4792] [Q&A] [answered]
I'm leading a project of introducing SPARQL as a query layer on top of some existing databases using RDF4J. For the time being, this will be proprietary software. As we understand the Eclipse license, we're allowed to use the RDF4J library in such a manner. I.e., we have dependencies on RDF4J maven artefacts and extends AbstractSail, AbstractSailConnection, etc.

### Custom optimiser pipeline
However, there is a point where things aren't so straight forward. E.g., we have a few custom query optimisers that detect certain query patterns that are handled in a particular way. So we want to reuse some of the pipeline from StandardQueryOptimizerPipeline but add and remove some optimisers. Squinting, this has similarities with how this was e.g. done [in Halyard](https://github.com/Merck/Halyard/blob/master/strategy/src/main/java/com/msd/gin/halyard/strategy/HalyardQueryOptimizerPipeline.java). Is this okay from a license point of view?

Very concretely, there is an overlap in our code:
```diff
11,12c11,12
< PROJECTION_REMOVAL_OPTIMIZER, // Make sure this is after the UnionScopeChangeOptimizer
< new QueryJoinOptimizer(evaluationStatistics, strategy.isTrackResultSize()),
---
> PROJECTION_REMOVAL_OPTIMIZER, // Make sure this is after the UnionScopeChangeOptimizer,
> new CustomOptimizer1(),
15c15,18
< ORDER_LIMIT_OPTIMIZER);
\ No newline at end of file
---
> ORDER_LIMIT_OPTIMIZER,
> new CustomOptimizer2(),
> ...
> new CustomOptimizerN()
\ No newline at end of file
```

### FilterOptimizer re-use
Another point of discussion are some customisations that we want to do based on `org.eclipse.rdf4j.query.algebra.evaluation.optimizer.FilterOptimizer`. Filters are in the way of some join optimisations that we're doing and essentially we'd like to re-use the `FilterUnMerger`, `FilterOrganizer` and `FilterMerger`, but can't without copying or other trickery because they're private (static) classes. Do you have any suggestions here?

**Comment by JervenBolleman** [comment, marked as answer]
The [license](https://github.com/eclipse-rdf4j/rdf4j/blob/main/LICENSE) question should be taken up with your own legal team. The license that RDF4j ships with is extremely liberal and allows many uses, but  the license does impose a few conditions and how to follow them in your special case/jurisdiction might need some thought. Thoughts that we, as developers, are not equipped to handle.

**Comment by JervenBolleman** [comment]
Regarding the FilterOptimizer changes, sure! Please create an issue and if you could make a pull request changing the permissions to public, I would appreciate it.

**Comment by frensjan** [comment]
Thanks Jerven.

About the FilterOptimizer, we were able to reuse it without de-aggregating its behaviour, so that issue seems of the table for now.
